README.md
[![Build Status](https://img.shields.io/travis/vancluever/aws-runas.svg)](https://travis-ci.org/vancluever/aws-runas)
[![Gem Version](https://img.shields.io/gem/v/aws_runas.svg)](https://rubygems.org/gems/aws_runas)
[![Codecov](https://img.shields.io/codecov/c/github/vancluever/aws-runas.svg)](https://codecov.io/github/vancluever/aws-runas)

aws-runas
==========

[![asciicast](https://asciinema.org/a/107502.png)](https://asciinema.org/a/107502)

**The problem:** You manage AWS across several different roles and need to use
tools outside of the regular `aws-cli` toolchain.

**The solution:** Use `aws-runas` :)

Features
---------

There are a few other tools and shell scripts out there that do the same
thing, but there are some differentiators in this gem:

 * Support for your roles already laid out in `~/.aws/config`.
  * These roles can also be copied to a local `aws_config` file and pushed
    to source control to ensure your deployment targets live with source.
 * Support for single-run commands (supplied on the command line) or
   interactive shell sessions (by supplying no commands).
 * MFA will be auto-detected and only prompted for if necessary (allowing one
   to assume a role that does not have a MFA serial supplied).
 * Session tokens can be acquired without assuming a role by adding the
   appropriate `mfa_serial` into the `[default]` profile and running `aws-runas`
   with `--no-role`. Subsequent uses of `aws-runas` after this will not prompt
   you for MFA (useful for tooling that needs to assume multiple roles off the
   same session token).
 * The session duration can be controlled with the `--duration` parameter. This
   allows you to change the session expiration if you require a period longer or
   shorter than an hour. The actual ranges you can choose with this setting
   depend on the account you are using and any configured maximums set on your
   role. More details can be found in the API documentation for
   [GetSesionToken][get-session-token] and [AssumeRole][assume-role].

[get-session-token]: https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html
[assume-role]: https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html

How it Works
-------------

Roles are assumed, or session tokens are simply acquired (if `--no-role` is
specified) via the [`AssumeRole`][assume-role] or the
[`GetSessionToken`][get-session-token] AWS STS API calls.  After this, your
command or shell is launched with the standard AWS credential chain environment
variables set:

 * `AWS_ACCESS_KEY_ID`
 * `AWS_SECRET_ACCESS_KEY`
 * `AWS_SESSION_TOKEN`

### Additional Variables

In addition to the above, the following environment variables are set to help
you gather additional information about the role and environment you are running
under:

 * `AWS_RUNAS_ASSUMED_ROLE_ARN` - set when a role is assumed (not set if
   `--no-role` is used)
 * `AWS_ROLE_SESSION_NAME` - contains the assumed role's session name (not set
   if `--no-role` is used). The format is
   `aws-runas-session_ACCTID_USERNAME_TIMESTAMP` when the user has access to
   [`GetCallerIdentity`][get-caller-identity], and `aws-runas-session_TIMESTAMP`
   format when they do not.
 * `AWS_RUNAS_PROFILE` - set with the profile used when `aws-runas` was run.
 * `AWS_REGION` and `AWS_DEFAULT_REGION` - set with the region name defined in
   the profile being used.
 * `AWS_SESSION_EXPIRATION` - set with the expiry timestamp in UTC.
 * `AWS_SESSION_EXPIRATION_UNIX` - set with the expiry timestamp in Unix time.

[get-caller-identity]: https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html

### Fancy Bash/Zsh Prompt

If you use `aws-runas` without any options and your default shell is Bash or
Zsh, a colorized prompt will appear with the profile that is in use if a role is
assumed, or a simple `(AWS)` indicator added to the prompt if a session token is
only obtained. See the video at the start of the doc for a demo!

#### Shell Integration Functions

2 functions currently get exported when you run under one of the two supported
shells:

 * `aws_session_expired`, which reads `AWS_SESSION_EXPIRATION_UNIX` (see above)
   and compares this with the current Unix timestamp supplied by `date`. It
   returns 0 on true and 1 on false, which can be used semantically in shell
   scripts.
 * `aws_session_status_color`, which works off of `aws_session_expired` to
   render either an ANSI color number (for bash)
   or a human readable color name (for zsh)
   - (red or 31) when `aws_session_expired` is `true`, (yellow or 33) otherwise.

#### Skipping the Fancy Prompt

If you are doing your own prompt customization based on aws-runas data and don't
want the prompt modified, you can supply `--skip-prompt` to skip the prompt
modification. The aforementioned integration functions will still be available
to you however, which you can use in your own scripts.

Usage
------

Install the gem (`gem install aws_runas`), and the command can be run via
`aws-runas` via your regular `$PATH`.

```
aws-runas: Run commands under AWS IAM roles

Usage:
  aws-runas [options] COMMAND ARGS

If COMMAND is omitted, the default shell ($SHELL, /bin/sh, or cmd.exe,
depending on your system) will launch.

[options] are:
  -n, --no-role         Get a session token only, do not assume a role
  -s, --skip-prompt     Do not launch interactive sessions with the fancy prompt
  -p, --path=<s>        Path to the AWS config file
  -r, --profile=<s>     The AWS profile to load (default: default)
  -d, --duration=<i>    The duration in seconds for temporary credentials (default: 3600)
  -v, --version         Print version and exit
  -h, --help            Show this message
```

`--path` is optional, and if omitted will default to the files in the
following order:

 * `aws_config`, in the current working directory
 * `~/.aws/config`, in your user directory.


Usage on Windows
-----------------

`aws_runas` works on Windows platforms, but YMMV. The gem has been tested
lightly on Cygwin and MSYS. Cygwin works great if you use the self-contained
Ruby ecosystem. Operating on MSYS or bare Windows will probably work as well as
any other Ruby gem. Running on WSL has not been tested, but as long as you can
get the minimum required Ruby version on it (currently >= 2.2.6), it should
generally work.

### OpenSSL Cert Bundle for Windows

Running `aws-runas` on native Windows may require the installation of a CA
certificate bundle. To do this, you will need to get the certificate bundle from
somewhere like [here](http://curl.haxx.se/docs/caextract.html) and set your
`SSL_CERT_FILE` environment variable to go to the file.


Author
-------

Chris Marchesi <chrism@vancluevertech.com>

License
--------

```
Copyright 2015-2018 Chris Marchesi

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
## v0.7.1

* Very minor update to address development dependency security issue
  (CVE-2020-8130 for Rake).

## v0.7.0

* Version now can be printed by supplying --version.
* The gem now uses [Optimist][ref-optimist], and as such should no longer give
  deprecation warnings for its previous name.
  [#21](https://github.com/vancluever/aws-runas/issues/21)
* Corrected an issue with session ID generation when the calculated new
  long-from session ID exceeded 64 characters. In this situation, the session
  name will fall back to the classic generic timestamped ID.
  [#17](https://github.com/vancluever/aws-runas/issues/17)
* When calling from an assumed role, the session ID now takes on the name of the
  access key ID instead of the account ID and user name. This should help
  prevent length or session name nesting issues, while still making the session
  name useful. [#17](https://github.com/vancluever/aws-runas/issues/17)
* The access key ID-based session ID will also be used if the account and
  user-based session ID would exceed 64 characters under normal circumstances,
  ensuring that the classic ID is strictly a fallback in major edge cases where
  either is unusable or if the user has no access to GetCallerIdentity.
  [#17](https://github.com/vancluever/aws-runas/issues/17)

[ref-optimist]: https://github.com/ManageIQ/optimist

## v0.6.0

### Session Duration Support

This update brings the `--duration` flag, which allows you to control the
session duration for both session tokens and assumed roles. Note that the
maximum depends on what kind of user you are using (IAM versus root account),
whether or not you are assuming a role or not, and the maximum duration set on
any role that you are assuming. `aws-runas` may silently truncate the maximum if
you request it too high, although your session will be rejected when assuming a
role. For more details, see [GetSesionToken][get-session-token] and
[AssumeRole][assume-role] for more details.

[get-session-token]: https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html
[assume-role]: https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html

Thanks to the work done in
[#16](https://github.com/vancluever/aws-runas/pull/16) for this!

### Zsh `PROMPT` Support

Some issues were discovered where the zsh prompt support was not functioning
correctly when using shell functions. Colors were not rendering properly as
well. Thanks to the work in
[#14](https://github.com/vancluever/aws-runas/pull/14) for the fix on this!

### Better Session IDs

`aws-runas` will now expose the IAM user's identity information (account ID/user
name) and enter it in the session ID, when available. The new format is
`aws-runas-session_ACCTID_USERNAME_TIMESTAMP` when the user has access to
[`GetCallerIdentity`][get-caller-identity], and the old
`aws-runas-session_TIMESTAMP` format when they do not.

Thanks to the work done in
[#11](https://github.com/vancluever/aws-runas/pull/11) for this!

[get-caller-identity]: https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html

## v0.5.0

### Zsh Support

`zsh` is now supported for the fancy prompt. When using the shell, your
existing profile data from `.zshrc` will be copied over into the temporary
configuration.

### Additional Functionality for Bash and Zsh

2 additional functions are available for `bash` and `zsh` as well:

 * `aws_session_expired`, which reads `AWS_SESSION_EXPIRATION_UNIX` (see below)
   and compares this with the current Unix timestamp supplied by `date`. It
   returns 0 on true and 1 on false, which can be used semantically in shell
   scripts.
 * `aws_session_status_color`, which works off of `aws_session_expired` to
   render an ANSI numeric color code - red when `aws_session_expired` is `true`,
   yellow otherwise.

The prompts for `bash` and `zsh` now work off of these functions to render the
right color. The prompt will go red when the session has expired.

### Skip Fancy Interactive Prompt

The default interactive prompt that you get when you run `aws-runas` with no
command supplied can now by skipped by adding `--skip-prompt` to the CLI
arguments. The profile functions mentioned above are still passed in. This
allows you to leverage their functionality inside your own scripts and custom
prompts if you want in other ways.

### Additional Variables

Several environment variables have been added for more quality-of-life when
working in the shell or aware tools:

 * `AWS_REGION` and `AWS_DEFAULT_REGION`, which pass through the region
   configured in the profile, if present
 * `AWS_SESSION_EXPIRATION` and `AWS_SESSION_EXPIRATION_UNIX` to supply the
   session expiration time, in both human and UNIX timestamps, respectively
   named.

## v0.4.2

The role that aws-runas assumed and the profile it used are now exposed as
`AWS_RUNAS_ASSUMED_ROLE_ARN` and `AWS_RUNAS_PROFILE`, respectively. These can be
used in scripts to track the profile being used or the role ARN used, in case
this data is needed later, or for troubleshooting purposes.

## v0.4.1

Fixed the escape sequence in the bash shell prompt indicator so that it has the
`\[` and `\]` enclosures - this fixes issues that the prompt was having with
line wrapping.

## v0.4.0

 * Dropping support for Ruby 2.1. You will need at least Ruby 2.2.6 to be using
   this gem now. If you have a version below this, please use a v0.3.x version.
 * MFA entry is no longer hidden from the terminal - you will see the digits you
   enter now.
 * Added a special indicator to `bash` prompts when running interactively. This
   prompt displays your running profile, like so: `(AWS:default)`. When running
   via --no-role, the indicator is just `(AWS)`. This should help to distinguish
   any AWS shells you may be running from regular ones.

## v0.3.1

This update sets `AWS_SDK_CONFIG_OPT_OUT` before the `aws-sdk` Ruby gem is
loaded to start assuming roles, to disable newer AWS Ruby SDK functionality that

spec/spec_helper.rb
# Copyright 2015 Chris Marchesi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'simplecov'
require 'codecov'
SimpleCov.start do
  add_filter '/vendor/'
  add_filter '/spec/'
end

SimpleCov.formatter = SimpleCov::Formatter::Codecov if ENV['CI'] == 'true'

require 'aws-sdk'

RSpec.configure do |config|
  config.color = true

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end
end

Aws.config.update(stub_responses: true)


spec/aws_runas/cli_spec.rb
# Copyright 2015 Chris Marchesi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'spec_helper'
require 'aws_runas/cli'
require 'aws_runas/main'

describe AwsRunAs::Cli do
  describe '::load_opts' do
    it 'loads the path option' do
      opts = AwsRunAs::Cli.load_opts(args: ['--path', 'test-opts/aws_config'])
      expect(opts[:path]).to eq('test-opts/aws_config')
    end

    it 'loads the profile option' do
      opts = AwsRunAs::Cli.load_opts(args: ['--profile', 'test-profile'])
      expect(opts[:profile]).to eq('test-profile')
    end
  end

  describe '::start' do
    before(:example) do
      allow(AwsRunAs::Cli).to receive(:load_opts).and_return({})
      allow(AwsRunAs::Cli).to receive(:read_mfa_if_needed)
      allow(AwsRunAs::Main).to receive(:new).and_return double(
        'AwsRunAs::Main',
        assume_role: true,
        handoff: true
      )
    end

    it 'creates an AwsConfig::Main instance' do
      expect(AwsRunAs::Main).to receive(:new)
      AwsRunAs::Cli.start
    end
  end

  describe '::read_mfa_if_needed' do
    it 'reads the MFA code' do
      allow(STDIN).to receive(:gets).and_return('123456')
      mfa_code = AwsRunAs::Cli.read_mfa_if_needed(
        path: MOCK_AWS_CONFIGPATH,
        profile: 'test-profile'
      )
      expect(mfa_code).to eq('123456')
    end
  end
end


spec/helpers/utils_spec.rb
require 'spec_helper'
require 'tmpdir'

MOCK_BASHRC_PATH = File.expand_path('../files/bashrc', __FILE__)

BASHRC_FILE_CONTENTS = <<EOS.freeze
foobar
EOS

ZSHRC_FILE_CONTENTS = <<EOS.freeze
bazqux
EOS

BASHRC_EXPECTED_PROMPT   = "PS1=\"\\[\\e[\\$(aws_session_status_color \"bash\")m\\](AWS:rspec)\\[\\e[0m\\] $PS1\"\n".freeze
ZSHRC_EXPECTED_PROMPT    = "PROMPT=\"%F{$(aws_session_status_color \"zsh\")}(AWS:rspec)%f $OLDPROMPT\"\n".freeze
ZSHRC_EXPECTED_SETSUBST  = "setopt PROMPT_SUBST\n".freeze
ZSHRC_EXPECTED_OLDPROMPT = "OLDPROMPT=\"$PROMPT\"\n".freeze
ZSH_MOCK_TMPDIR          = "#{Dir.tmpdir}/aws_runas_zsh_rspec".freeze

EXPECTED_ENV = {
  'AWS_ACCESS_KEY_ID' => 'AccessKeyId',
  'AWS_SECRET_ACCESS_KEY' => 'SecretAccessKey',
  'AWS_SESSION_TOKEN' => 'Token'
}.freeze

EXPECTED_ENV_ZSH = {
  'AWS_ACCESS_KEY_ID' => 'AccessKeyId',
  'AWS_SECRET_ACCESS_KEY' => 'SecretAccessKey',
  'AWS_SESSION_TOKEN' => 'Token',
  'ZDOTDIR' => ZSH_MOCK_TMPDIR
}.freeze

def test_mktmpdir
  Dir.mkdir(ZSH_MOCK_TMPDIR)
  ZSH_MOCK_TMPDIR
end


lib/aws_runas/cli.rb
# Copyright 2015 Chris Marchesi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'optimist'
require 'aws_runas/config'
require 'aws_runas/main'
require 'aws_runas/version'
require 'io/console'

module AwsRunAs
  module Cli
    module_function

    # loads the command-line options.
    def load_opts(args: ARGV)
      Optimist.options(args) do
        version AwsRunAs::VERSION
        banner <<-EOS.gsub(/^ {10}/, '')
          aws-runas: Run commands under AWS IAM roles

          Usage:
            aws-runas [options] COMMAND ARGS

          If COMMAND is omitted, the default shell ($SHELL, /bin/sh, or cmd.exe,
          depending on your system) will launch.

          [options] are:
        EOS

        opt :no_role, 'Get a session token only, do not assume a role', type: TrueClass, default: nil
        opt :skip_prompt, 'Do not launch interactive sessions with the fancy prompt', type: TrueClass, default: nil
        opt :path, 'Path to the AWS config file', type: String
        opt :profile, 'The AWS profile to load', type: String, default: 'default'
        opt :duration, 'The duration in seconds for temporary credentials', type: Integer, default: 3600
        stop_on_unknown
      end
    end

    # Start the CLI. Load options (profile, specific config path), and run
    # main.
    def start
      opts = load_opts
      mfa_code = read_mfa_if_needed(path: opts[:path], profile: opts[:profile])
      @main = AwsRunAs::Main.new(path: opts[:path], profile: opts[:profile], mfa_code: mfa_code, no_role: opts[:no_role], duration_seconds: opts[:duration])
      @main.assume_role
      command = ARGV.shift
      @main.handoff(command: command, argv: ARGV, skip_prompt: opts[:skip_prompt])
    end

    # Reads the MFA code from standard input.
    def read_mfa_if_needed(path: nil, profile: 'default')
      @cfg = AwsRunAs::Config.new(path: path, profile: profile)
      return nil unless @cfg.mfa_required?
      STDOUT.print 'Enter MFA code: '
      STDOUT.flush
      STDIN.gets.chomp
    end
  end
end


lib/aws_runas/utils.rb
# Copyright 2016 Chris Marchesi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'rbconfig'
require 'tempfile'
require 'tmpdir'
require 'fileutils'
require 'English'

module AwsRunAs
  # Utility functions that aren't specifically tied to a class.
  module Utils
    module_function

    # Return the path to the shell_profiles directory vendored with the gem.
    def shell_profiles_dir
      File.expand_path('../../../shell_profiles', __FILE__)
    end

    # Run an interactive bash session with a special streamed RC file.  The RC
    # merges a local .bashrc if it exists, with a prompt that includes the
    # computed message from handoff_to_shell.
    def handoff_bash(env:, path:, message:, skip_prompt:)
      rc_data = IO.read("#{ENV['HOME']}/.bashrc") if File.exist?("#{ENV['HOME']}/.bashrc")
      rc_file = Tempfile.new('aws_runas_bashrc')
      rc_file.write("#{rc_data}\n") unless rc_data.nil?
      rc_file.write(IO.read("#{shell_profiles_dir}/sh.profile"))
      unless skip_prompt
        rc_file.write("PS1=\"\\[\\e[\\$(aws_session_status_color \"bash\")m\\](#{message})\\[\\e[0m\\] $PS1\"\n")
      end
      rc_file.close
      system(env, path, '--rcfile', rc_file.path)
    ensure
      rc_file.unlink
    end

    # Run an interactive zsh session with a special streamed RC file.  The RC
    # merges a local .zshrc if it exists, with a prompt that includes the
    # computed message from handoff_to_shell.
    def handoff_zsh(env:, path:, message:, skip_prompt:)
      rc_data = IO.read("#{ENV['HOME']}/.zshrc") if File.exist?("#{ENV['HOME']}/.zshrc")
      rc_dir = Dir.mktmpdir('aws_runas_zsh')
      rc_file = File.new("#{rc_dir}/.zshrc", 'w')
      rc_file.write("#{rc_data}\n") unless rc_data.nil?
      rc_file.write(IO.read("#{shell_profiles_dir}/sh.profile"))
      unless skip_prompt
        rc_file.write("setopt PROMPT_SUBST\n")
        rc_file.write("OLDPROMPT=\"$PROMPT\"\n")
        rc_file.write("PROMPT=\"%F{$(aws_session_status_color \"zsh\")}(#{message})%f $OLDPROMPT\"\n")
      end
      rc_file.close
      env.store('ZDOTDIR', rc_dir)
      system(env, path)
    ensure
      FileUtils.rmtree(rc_dir)
    end

    # load the shell for a specific operating system.
    # if $SHELL exists, load that.
    def shell
      if RbConfig::CONFIG['host_os'] =~ /mswin|windows|mingw32/i
        'cmd.exe'
      elsif ENV.include?('SHELL')
        ENV['SHELL']
      else
        '/bin/sh'
      end
    end

    # Compute the message given to the prompt based off supplied profile.
    def compute_message(profile:)
      if profile.nil?
        'AWS'
      else
        "AWS:#{profile}"
      end
    end

    # "Handoff" to a supported interactive shell. More technically, this runs
    # an interactive shell with the shell prompt customized to the current
    # running AWS profile. If the shell is not something we can handle
    # specifically, just run the shell.
    def handoff_to_shell(env:, profile: nil, skip_prompt:)
      path = shell
      if path.end_with?('/bash')
        handoff_bash(env: env, path: path, message: compute_message(profile: profile), skip_prompt: skip_prompt)
      elsif path.end_with?('/zsh')
        handoff_zsh(env: env, path: path, message: compute_message(profile: profile), skip_prompt: skip_prompt)
      else
        system(env, path)
      end
      exit $CHILD_STATUS.exitstatus
    end
  end
end


lib/aws_runas/config.rb
# Copyright 2015 Chris Marchesi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'inifile'

module AwsRunAs
  # Manages the configuartion file, including loading and retrieving values.
  class Config
    attr_reader :profile
    # Finds the configuration file (used if no file is specified).
    # paths searched: ./aws_config, and ~/.aws/config.
    def self.find_config_file
      local_config = File.expand_path('aws_config')
      user_config = File.expand_path('~/.aws/config')
      return local_config if File.exist?(local_config)
      user_config if File.exist?(user_config)
    end

    def initialize(path:, profile:)
      @path = path
      @path = self.class.find_config_file unless @path
      fail(Errno::ENOENT, "#{@path}") unless File.exist?(@path.to_s)
      @profile = profile
    end

    # Loads the config section for a specific profile.
    def load_config_value(key:)
      section = @profile
      section = "profile #{@profile}" unless @profile == 'default'
      aws_config = IniFile.load(@path)
      unless aws_config.has_section?(section)
        fail(NameError, "Profile #{@profile} not found in #{@path}")
      end
      aws_config[section][key]
    end

    # Checks to see if MFA is required for a specific profile.
    def mfa_required?
      return true if load_config_value(key: 'mfa_serial') && !ENV.include?('AWS_SESSION_TOKEN')
      false
    end

    # loads the soruce credentials profile based on the supplied profile.
    def load_source_profile
      source_profile = load_config_value(key: 'source_profile')
      return source_profile if source_profile
      @profile
    end
  end
end


lib/aws_runas/main.rb
# Copyright 2015 Chris Marchesi
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'aws_runas/config'
require 'aws_runas/utils'

# AWS_SDK_CONFIG_OPT_OUT must be set here so that we use the pre-2.4 SDK
# behaviour, which ensures that ~/.aws/config is not re-read when assuming
# roles.
ENV.store('AWS_SDK_CONFIG_OPT_OUT', '1')
require 'aws-sdk'

module AwsRunAs
  # Main program logic for aws-runas - sets up sts asession and assumed role,
  # and hands off environment to called process.
  class Main
    # Instantiate the object and set up the path, profile, and populate MFA.
    #
    # Session timestamp is also registered here to prevent races.
    def initialize(path: nil, profile: default, mfa_code: nil, no_role: nil, duration_seconds: 3600)
      cfg_path = if path
                   path
                 else
                   AwsRunAs::Config.find_config_file
                 end
      @cfg = AwsRunAs::Config.new(path: cfg_path, profile: profile)
      @mfa_code = mfa_code
      @no_role = no_role
      @duration_seconds = duration_seconds
      @session_timestamp = Time.now.to_i
    end

    def sts_client
      region = @cfg.load_config_value(key: 'region')
      region = 'us-east-1' unless region
      Aws::STS::Client.new(
        profile: @cfg.load_source_profile,
        region: region
      )
    end

    def render_classic_session
      "aws-runas-session_#{@session_timestamp}"
    end

    def render_account_user_session(caller_identity)
      "aws-runas-session_#{caller_identity.account}_#{caller_identity.arn.split('/')[-1]}_#{@session_timestamp}"
    end

    def render_access_key_session(caller_identity)
      "aws-runas-session_#{caller_identity.user_id.split(':')[0]}_#{@session_timestamp}"
    end

    def use_access_key_id?(caller_identity)
      caller_identity.arn =~ %r{^arn:aws:sts::\d+:assumed-role\/} ||
        render_account_user_session(caller_identity).length > 64
    end

    # Returns a session name based on the following criteria:
    #  * If the user ARN matches that of an assumed role, return the access key ID
    #  * Otherwise, return the account ID and the user (last part of the ARN).
    def render_modern_session(caller_identity)
      return render_access_key_session(caller_identity) if use_access_key_id?(caller_identity)
      render_account_user_session(caller_identity)
    end

    def session_id
      caller_identity = sts_client.get_caller_identity
      return render_classic_session if render_modern_session(caller_identity).length > 64
      render_modern_session(caller_identity)
    rescue Aws::STS::Errors::AccessDeniedException
      render_classic_session
    end

    def assume_role
      @role_session_name = session_id
      role_arn = @cfg.load_config_value(key: 'role_arn')
      mfa_serial = @cfg.load_config_value(key: 'mfa_serial') unless ENV.include?('AWS_SESSION_TOKEN')
      if @no_role
        raise 'No mfa_serial in selected profile, session will be useless' if mfa_serial.nil?
        @session = sts_client.get_session_token(
          serial_number: mfa_serial,
          token_code: @mfa_code,
          duration_seconds: @duration_seconds
        )
      else
        @session = Aws::AssumeRoleCredentials.new(
          client: sts_client,
          role_arn: role_arn,
          serial_number: mfa_serial,
          token_code: @mfa_code,
          role_session_name: @role_session_name,
          duration_seconds: @duration_seconds
        )
      end
    end

    def session_credentials
      @session.credentials
    end

    def credentials_env
      env = {}
      env['AWS_ACCESS_KEY_ID'] = session_credentials.access_key_id
      env['AWS_SECRET_ACCESS_KEY'] = session_credentials.secret_access_key
      env['AWS_SESSION_TOKEN'] = session_credentials.session_token
      env['AWS_RUNAS_PROFILE'] = @cfg.profile
      unless @cfg.load_config_value(key: 'region').nil?
        env['AWS_REGION'] = @cfg.load_config_value(key: 'region')
        env['AWS_DEFAULT_REGION'] = @cfg.load_config_value(key: 'region')
      end
      if @no_role
        env['AWS_SESSION_EXPIRATION'] = session_credentials.expiration.to_s
        env['AWS_SESSION_EXPIRATION_UNIX'] = DateTime.parse(session_credentials.expiration.to_s).strftime('%s')
      else
        env['AWS_SESSION_EXPIRATION'] = @session.expiration.to_s
        env['AWS_SESSION_EXPIRATION_UNIX'] = DateTime.parse(@session.expiration.to_s).strftime('%s')
        env['AWS_RUNAS_ASSUMED_ROLE_ARN'] = @cfg.load_config_value(key: 'role_arn')
        env['AWS_ROLE_SESSION_NAME'] = @role_session_name
      end
      env
    end

    def handoff(command: nil, argv: nil, skip_prompt:)
      env = credentials_env
      unless command
        AwsRunAs::Utils.handoff_to_shell(env: env, profile: @no_role ? nil : @cfg.profile, skip_prompt: skip_prompt)
      end
      exec(env, command, *argv)
    end
  end
end
